### PR TITLE
add model-as-a-service documentation

### PR DESCRIPTION
**Pull request by ImenBizid**

### CODE REVIEW

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 1]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
+|===
+| *Variable name* | *Description* | *Type*
+| `DOCKER_IMAGE`
+| the Docker image that contains all the dependencies and libraries needed for the model execution.
+| String (default="activeeon/dlm3")
+| `INSTANCE_NAME`
+| the name of the service that will be deployed.
+| String (default="model-server")
+| `YAML_FILE`
+| A YAML file that describes the OpenAPI Specification (known as Swagger Spec) of the service. This file should be stored in the catalog under the `model_as_service_resources` bucket. This file should be stored in the Catalog. More information about the structure of this file can be found in the section <<Customize the Service>>.
```
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: remove this sentence: `This file should be stored in the Catalog.`

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 2]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
+Search for `Model_Service` in Services Workflows List.
+
+Set the following variables :
+
+.Model_Service variables
+[cols="2,5,2"]
+|===
+| *Variable name* | *Description* | *Type*
+| `DOCKER_IMAGE`
+| the Docker image that contains all the dependencies and libraries needed for the model execution.
```
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: `The` instead of `the`

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 3]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
+| String (default="activeeon/dlm3")
+| `INSTANCE_NAME`
+| the name of the service that will be deployed.
```
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: `The` instead of `the`

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 4]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
+| String (default="model-server")
+| `YAML_FILE`
+| A YAML file that describes the OpenAPI Specification (known as Swagger Spec) of the service. This file should be stored in the catalog under the `model_as_service_resources` bucket. This file should be stored in the Catalog. More information about the structure of this file can be found in the section <<Customize the Service>>.
+| String (default="ml_service-api").
+| `PYTHON_ENTRYPOINT`
+| This entry script starts the service and defines the different functions to deploy the model, score the requests by using the model, and return the results. This specific script is specific to your model. This file should be stored in the Catalog. More information about this file can be found in the <<Customize the Service>> section.
```
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: `This specific script is specific to your model.` => `This script is specific to your model.`

File: src/docs/MLOS/MLOSUserGuide.adoc (hunk 1 above)
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: Remove this sentence: `This file should be stored in the Catalog.`

File: src/docs/MLOS/MLOSUserGuide.adoc (hunk 2 above)
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: `The` instead of `the`

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 5]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
-//| `DOCKER_IMAGE`
-//| Specifies the docker image that will be used to execute the task.
-//| String (default="activeeon/dlm3")
-//| `INPUT_VARIABLES`
-//| Specifies the x and y values.
-//| JSON format (default={"x": 3.5976705887960527, "y": -1.8471754188451843})
-//| `TOKEN`
-//|  Specifies the id for the tested parameters.
-//| String (default=-1)
-//|===
```
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: Why this block was `commented` and then `removed` ?

File: src/docs/MLOS/MLOSUserGuide.adoc (hunk 1 above)
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: `OpenAPI Specification ver. 2`

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 6]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
+
+Open the link:https://try.activeeon.com/Studio[Studio Portal].
+
+Create a new workflow.
+
+Add the `model_as_a_service` bucket.
+
+Drag and drop the `Model_Service_Start` task from the bucket.
+
+Execute the Workflow by setting the different workflow variables as following:
```
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: `as following` => `as follows`
https://ell.stackexchange.com/questions/2747/as-following-vs-as-follows

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 7]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
+
+.Model_Service_Start variables
+[cols="2,5,2"]
+|===
+| *Variable name* | *Description* | *Type*
+| `DOCKER_IMAGE`
+| the Docker image that contains all the dependencies and libraries needed for the model execution.
```
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: `The` instead of `the`

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 8]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
+| String (default="activeeon/dlm3")
+| `INSTANCE_NAME`
+| the name of the service that will be deployed.
```
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: `The` instead of `the`

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 9]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
+| String (default="model-server")
+| `YAML_FILE`
+| A YAML file that describes the OpenAPI Specification (known as Swagger Spec) of the service. This file should be stored in the Catalog under the `model_as_service_resources` bucket. More information about the structure of this file can be found in the section <<Customize the Service>>.
```
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: `OpenAPI Specification ver. 2`

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 10]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
+| String (default="ml_service-api").
+| `PYTHON_ENTRYPOINT`
+| This entry script starts the service and defines the different functions to deploy the model, score the requests by using the model, and return the results. This specific script is specific to your model. This file should be stored in the Catalog under the `model_as_service_resources` bucket. More information about this file can be found in the <<Customize the Service>> section.
```
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: `This specific script is specific to your model.` . => `This script is specific to your model.`

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 11]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
+|===
+
+
+=== Deploy a Model
+
+There is two ways to deploy a model in the started model_server (i.e. where you will deploy, update and call your model) using MLOS:
+
+*Via Swagger UI*
+
+Open the swagger UI.
```
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: `swagger` => `Swagger`

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 12]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
+
+Open the link:https://try.activeeon.com/Studio[Studio Portal].
+
+Create a new workflow.
+
+Add the `model_as_a_service` bucket.
+
+Drag and drop the `Deploy_Model` task from the bucket.
+
+Execute the Workflow and set the different workflow variables as following:
```
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: `as following` => `as follows`
https://ell.stackexchange.com/questions/2747/as-following-vs-as-follows

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 13]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
+
+.Deploy_Model variables
+[cols="2,5,2"]
+|===
+| *Variable name* | *Description* | *Type*
+| `DEPLOY_MODEL_ENDPOINT`
+| the endpoint of the started service.
+| URL
+| `API_EXTENSION`
+| the base path to access the deployment endpoint.
```
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: `the` => `The`

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 14]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
+| String (default="/api/deploy")
+| `MODEL_URL`
+| the URL of the model that need to be deployed.
```
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: `the` => `The`

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 15]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
+| URL
+| `SERVICE_TOKEN`
+| the api token to be able to deploy the model.
```
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: `the` => `The`

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 16]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
+|===
+
+
+=== Call the service for predictions
+
+There is two ways to call the web service to score input data using MLOS:
+
+*Via Swagger UI*
+
+Open the swagger UI.
```
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: `swagger` => `Swagger`

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 17]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
+
+Open the link:https://try.activeeon.com/Studio[Studio Portal].
+
+Create a new workflow.
+
+Add the `model_as_a_service` bucket.
+
+Drag and drop the `Call_Prediction_Service` task from the bucket.
+
+Execute the Workflow and set the different workflow variables as following:
```
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: `as following` => `as follows`
https://ell.stackexchange.com/questions/2747/as-following-vs-as-follows

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 18]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
+
+.Call_Prediction_Service variables
+[cols="2,5,2"]
+|===
+| *Variable name* | *Description* | *Type*
+| `PREDICT_MODEL_ENDPOINT`
+| the endpoint of the started service.
+| URL
+| `PREDICT_EXTENSION`
+| the base path to access the prediction endpoint.
```
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: `the` => `The`

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 19]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
+| String (default="/api/predict")
+| `SERVICE_TOKEN`
+| the api token to be able to deploy the model.
```
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: `the` => `The`

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 20]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
+| String
+| `INPUT_DATA`
+| entry data that needs to be scored by the deployed model.
```
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: `entry` => `Entry` or `Input`

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 21]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
+| JSON
+| `LABEL_COLUMN`
+| need to be set if data is labeled.
```
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: `need to be set if data is labeled.` => `Name of the label column. Need to be set if data is labeled.`

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 22]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
+|===
+| *Variable name* | *Description* | *Type*
+| `ACTION`
+| The action that will be processed regarding the service status.
+| List (Pause_Model_Service, Resume_Model_Service, Finish_Model_Service) (default="Finish_Model_Service")
+| `INSTANCE_NAME`
+| the service name that need to be deleted, resumed or paused.
+| String (default="model-server-${PA_JOB_ID}")
+| `INSTANCE_ID`
+| the service instance id
```
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: `the` => `The`

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 23]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
+| String
+|===
+
+=== Deployment Pipeline Example
+
+You can connect the different tasks in a single workflow to get the full pipeline from the model training step to the model deployment and consumption steps. Each task will propagate the acquired variables to the connected tasks.
```
Comment (andrewssobral) [inline src/docs/MLOS/MLOSUserGuide.adoc]: `acquired` => `given`

File: src/docs/MLOS/MLOSUserGuide.adoc (hunk 5 above)
Comment (ImenBizid) [reply src/docs/MLOS/MLOSUserGuide.adoc]: It is commented as for the other objective examples workflows, we didn't provide a detailed description for the variables. So, we have done the same for Himmelblau to keep the documentation lite and homogeneous.

File: src/docs/MLOS/MLOSUserGuide.adoc (hunk 23 above)
Comment (ImenBizid) [reply src/docs/MLOS/MLOSUserGuide.adoc]: It is more acquired than given. As I mean the resulted variables. I can use resulted to be more clear.

[the remark this answers is not in the document]
Comment (ImenBizid) [reply src/docs/MLOS/MLOSUserGuide.adoc]: from the model training step to the model deployment and consumption steps.
from step to step. that's why !

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 24]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
+The following workflow example is available on the `model_as_a_service` bucket under the name of `Deploy_And_Call_IRIS_Flower_Classifier_Model`.
+This example trains an Iris Flower Classifier, starts a web service where the trained model is deployed and input data is scored by consuming the endpoints exposed by the model_service.
```
[the remark this answers is not in the document]
Comment (ImenBizid) [reply src/docs/MLOS/MLOSUserGuide.adoc]: ok

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 25]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
+In the following, we describe in depth the content of each element :
+
+*PYTHON_ENTRYPOINT file* :  The following python script refers to the ml_service.py file stored in the catalog under the `model_as_a_service_resources` bucket.
+This script defines the different functions needed to deploy the model, score data and generate tokens.
+It is possible to edit this script to make it more customized to your model. The entry script must take into consideration:
+
+* the list of users that are allowed to consume the service endpoints
+* the format of the model expected by the deployment and the prediction functions
+* the format of the incoming data
+* the data format expected by the model
```
[the remark this answers is not in the document]
Comment (ImenBizid) [reply src/docs/MLOS/MLOSUserGuide.adoc]: ok

File: src/docs/MLOS/MLOSUserGuide.adoc [hunk 26]
```diff
@@ -331,24 +331,493 @@ The following workflows represent some mathematical functions that can be optimi
...
+                type: string
+                default: "376d873c859d7f9f268e1b9be883745b"
+----
+<1> Contains API information: title, description (optional), version, licence.
+<2> Defines the get_token endpoint calling the get_token` function defined in the entry script.
+<3> Defines the deploy endpoint calling the `deploy_api`` function defined in the entry script.
+<4> Defines the predict endpoint calling the `predict_api` function defined in the entry script.
+--
+
+*DOCKER_IMAGE name* : Choose your own image containing the different dependencies required to run your ENTRYPOINT_SCRIPT.
```
[the remark this answers is not in the document]
Comment (ImenBizid) [reply src/docs/MLOS/MLOSUserGuide.adoc]: ok

File: src/docs/MLOS/MLOSUserGuide.adoc (hunk 1 above)
Comment (ImenBizid) [reply src/docs/MLOS/MLOSUserGuide.adoc]: ok

File: src/docs/MLOS/MLOSUserGuide.adoc (hunk 6 above)
Comment (ImenBizid) [reply src/docs/MLOS/MLOSUserGuide.adoc]: ok

File: src/docs/MLOS/MLOSUserGuide.adoc (hunk 7 above)
Comment (ImenBizid) [reply src/docs/MLOS/MLOSUserGuide.adoc]: I will replace all of them

File: src/docs/MLOS/MLOSUserGuide.adoc (hunk 10 above)
Comment (ImenBizid) [reply src/docs/MLOS/MLOSUserGuide.adoc]: ok

File: src/docs/MLOS/MLOSUserGuide.adoc (hunk 23 above)
Comment (andrewssobral) [reply src/docs/MLOS/MLOSUserGuide.adoc]: ok, let keep `acquired`

[the remark this answers is not in the document]
Comment (andrewssobral) [reply src/docs/MLOS/MLOSUserGuide.adoc]: ok

File: src/docs/MLOS/MLOSUserGuide.adoc (hunk 2 above)
Comment (andrewssobral) [reply src/docs/MLOS/MLOSUserGuide.adoc]: `the Docker image ` => `The docker image `

**Pull request merged**